presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java

```diff
@@ -0,0 +1,295 @@
+    private final List<Integer> partitionChannels;
+    private final List<Optional<Block>> partitionConstants;
+    private final boolean replicateNullsAndAny;
+    private final OptionalInt nullChannel;
+
+    private boolean finished;
+    private boolean hasAnyRowBeenReplicated;
+
+    public PrestoSparkOutputOperator(
+            OperatorContext operatorContext,
+            PrestoSparkRowBuffer rowBuffer,
+            Function<Page, Page> pagePreprocessor,
+            PartitionFunction partitionFunction,
+            List<Integer> partitionChannels,
+            List<Optional<Block>> partitionConstants,
+            boolean replicateNullsAndAny,
+            OptionalInt nullChannel)
+    {
+        this.operatorContext = requireNonNull(operatorContext, "operatorContext is null");
+        this.rowBuffer = requireNonNull(rowBuffer, "rowBuffer is null");
+        this.pagePreprocessor = requireNonNull(pagePreprocessor, "pagePreprocessor is null");
+        this.partitionFunction = requireNonNull(partitionFunction, "partitionFunction is null");
+        this.partitionChannels = ImmutableList.copyOf(requireNonNull(partitionChannels, "partitionChannels is null"));
+        this.partitionConstants = ImmutableList.copyOf(requireNonNull(partitionConstants, "partitionConstants is null"));
+        this.replicateNullsAndAny = replicateNullsAndAny;
+        this.nullChannel = requireNonNull(nullChannel, "nullChannel is null");
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        return rowBuffer.isFull();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished && isBlocked().isDone();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        page = pagePreprocessor.apply(page);
+        int positionCount = page.getPositionCount();
+        int channelCount = page.getChannelCount();
+        int averageRowSizeInBytes = min(toIntExact(page.getLogicalSizeInBytes() / positionCount), 10);
+        Page partitionFunctionArguments = getPartitionFunctionArguments(page);
+        for (int position = 0; position < positionCount; position++) {
+            SliceOutput output = new DynamicSliceOutput(averageRowSizeInBytes * 2);
+            for (int channel = 0; channel < channelCount; channel++) {
+                Block block = page.getBlock(channel);
+                block.writePositionTo(position, output);
+            }
+
+            boolean shouldReplicate = (replicateNullsAndAny && !hasAnyRowBeenReplicated) ||
+                    nullChannel.isPresent() && page.getBlock(nullChannel.getAsInt()).isNull(position);
+            if (shouldReplicate) {
+                for (int i = 0; i < partitionFunction.getPartitionCount(); i++) {
+                    rowBuffer.enqueue(new PrestoSparkRow(i, output.size(), output.getUnderlyingSlice().byteArray()));
```

Comment:
I guess you intend to avoid create a new `DynamicSliceOutput` each time so you try to get the underlying byte array directly 😃 . I understand it cannot be done for now since you cannot reset `DynamicSliceOutput` 

That being said, we might want to consider add something like `reset` into `SliceOutput`?

Reply:
This should be done as part of the buffer recycling effort. As SliceOutput has to acquire a free buffer from a pool before you can reset it.